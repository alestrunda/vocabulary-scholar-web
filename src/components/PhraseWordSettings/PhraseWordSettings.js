import React from 'react';
import PropTypes from 'prop-types';
import TextHighlight from '../../components/TextHighlight';
import QuestionPhrase from '../../components/QuestionPhrase';
import TextField from 'material-ui/TextField';
import RaisedButton from 'material-ui/RaisedButton';

class PhraseWordSettings extends React.Component {
  state = {
    phraseField: '',
    phraseFieldError: undefined,
    phraseHighlight: {},
  };

  componentDidMount() {
    if (this.props.phrase) this.loadStateFromProps();
  }

  componentDidUpdate(prevProps) {
    if (
      !prevProps.phrase ||
      prevProps.phrase.phrase !== this.props.phrase.phrase
    )
      this.loadStateFromProps();
  }

  loadStateFromProps() {
    if (!this.props.phrase) return;
    let wordPosition;
    if (
      this.props.phrase.startIndex !== undefined &&
      this.props.phrase.endIndex !== undefined
    ) {
      //get phrase's highlight indexes from props
      wordPosition = {
        start: this.props.phrase.startIndex,
        end: this.props.phrase.endIndex,
      };
    } else {
      //phrase's highlight indexes not set, try to locate current word in the phrase and use those indexes
      wordPosition = this.locateWordInPhrase(
        this.props.word,
        this.props.phrase.phrase
      );
    }
    this.setState({
      phraseField: this.props.phrase.phrase,
      phraseHighlight: {
        start: wordPosition ? wordPosition.start : undefined,
        end: wordPosition ? wordPosition.end : undefined,
      },
    });
  }

  resetSettings() {
    this.setState({
      phraseField: '',
      phraseFieldError: undefined,
      phraseHighlight: {},
    });
  }

  locateWordInPhrase(word, phrase) {
    const wordIndex = phrase.toLowerCase().indexOf(word.toLowerCase());
    if (wordIndex !== -1) {
      return {
        start: wordIndex,
        end: wordIndex + word.length,
      };
    }
  }

  handlePhraseFieldChange = event => {
    const newPhrase = event.target.value;
    let start, end;
    //if highlight not set yet, make highlight prediction - try to locate current word and highlight it
    if (this.state.phraseHighlight.start === undefined) {
      const wordPosition = this.locateWordInPhrase(this.props.word, newPhrase);
      if (wordPosition) {
        start = wordPosition.start;
        end = wordPosition.end;
      }
    } else {
      //because phrase has changed, make sure the highlight indexes woun't get outside the phrase length
      start =
        newPhrase.length < this.state.phraseHighlight.start
          ? newPhrase.length
          : this.state.phraseHighlight.start;
      end =
        newPhrase.length < this.state.phraseHighlight.end
          ? newPhrase.length
          : this.state.phraseHighlight.end;
    }
    this.setState({
      phraseField: newPhrase,
      phraseHighlight: {
        start,
        end,
      },
    });
  };

  handleSubmitButtonClick = () => {
    if (!this.state.phraseField) {
      this.setState({
        phraseFieldError: 'This field cannot be empty',
      });
      return;
    }
    if (this.state.phraseHighlight.start === this.state.phraseHighlight.end) {
      this.setState({
        submitError:
          'Please select some part of the phrase - during lesson this part will be hidden and the goal will be to put the phrase back together',
      });
      return;
    }
    this.props.onSubmit(
      this.state.phraseField,
      this.state.phraseHighlight.start,
      this.state.phraseHighlight.end
    );
    this.resetSettings();
  };

  handleOnPhraseHighlight = (start, end) => {
    this.setState({
      phraseHighlight: {
        start,
        end,
      },
      submitError: '',
    });
  };

  render() {
    return (
      <React.Fragment>
        <h3>1. Set phrase</h3>
        <TextField
          className="mb25"
          style={{
            width: '100%',
          }}
          floatingLabelText="Phrase"
          onChange={this.handlePhraseFieldChange}
          value={this.state.phraseField}
          errorText={this.state.phraseFieldError}
        />
        <h3 className="mb5">2. Select part that will be hidden</h3>
        <TextHighlight
          className="mb30"
          onHighlight={this.handleOnPhraseHighlight}
          start={this.state.phraseHighlight.start}
          end={this.state.phraseHighlight.end}
        >
          {this.state.phraseField}
        </TextHighlight>
        <h3 className="mb5">3. Question preview</h3>
        <QuestionPhrase
          word={this.props.word}
          wordID={this.props.wordID}
          phrase={this.state.phraseField}
          startIndex={this.state.phraseHighlight.start}
          endIndex={this.state.phraseHighlight.end}
          disabled
          hideIfNotAvailable
        />
        {this.state.submitError && (
          <p className="text-error mb10">{this.state.submitError}</p>
        )}
        <div className="mb25 mt35">
          <RaisedButton
            primary
            label={this.props.submitText}
            onClick={this.handleSubmitButtonClick}
          />
        </div>
      </React.Fragment>
    );
  }
}

PhraseWordSettings.propTypes = {
  wordID: PropTypes.string.isRequired,
  word: PropTypes.string.isRequired,
  submitText: PropTypes.string,
  phrase: PropTypes.object,
  onSubmit: PropTypes.func.isRequired,
};

PhraseWordSettings.defaultProps = {
  submitText: 'Save phrase',
};

export default PhraseWordSettings;
