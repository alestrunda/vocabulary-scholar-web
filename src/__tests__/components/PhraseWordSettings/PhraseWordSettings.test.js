import React from 'react';
import { shallow } from 'enzyme';

import PhraseWordSettings from '../../../components/PhraseWordSettings';

it('renders correctly', () => {
  const wordID = 'school',
    word = 'School',
    onSubmit = jest.fn();
  const root = shallow(
    <PhraseWordSettings wordID={wordID} word={word} onSubmit={onSubmit} />
  );
  expect(root).toMatchSnapshot();
});

it('renders correctly with phrase data passed', () => {
  const wordID = 'school',
    word = 'School',
    onSubmit = jest.fn();
  const phrase = {
    phrase: 'Lorem ipsum',
    startIndex: 0,
    endIndex: 10,
  };
  const root = shallow(
    <PhraseWordSettings
      wordID={wordID}
      word={word}
      phrase={phrase}
      onSubmit={onSubmit}
    />
  );
  expect(root).toMatchSnapshot();
});

it('triggers onSubmit', () => {
  const wordID = 'school',
    word = 'School',
    onSubmit = jest.fn();
  const phrase = {
    phrase: 'Lorem ipsum',
    startIndex: 0,
    endIndex: 10,
  };
  const root = shallow(
    <PhraseWordSettings
      wordID={wordID}
      word={word}
      phrase={phrase}
      onSubmit={onSubmit}
    />
  );
  root.find('RaisedButton').simulate('click');
  expect(onSubmit).toHaveBeenCalledWith(
    phrase.phrase,
    phrase.startIndex,
    phrase.endIndex
  );
});
