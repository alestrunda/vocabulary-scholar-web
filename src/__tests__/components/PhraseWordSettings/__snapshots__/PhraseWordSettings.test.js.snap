// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PhraseWordSettings
    onSubmit={[MockFunction]}
    submitText="Save phrase"
    word="School"
    wordID="school"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h3>
          1. Set phrase
        </h3>,
        <TextField
          className="mb25"
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelText="Phrase"
          fullWidth={false}
          multiLine={false}
          onChange={[Function]}
          rows={1}
          style={
            Object {
              "width": "100%",
            }
          }
          type="text"
          underlineShow={true}
          value=""
        />,
        <h3
          className="mb5"
        >
          2. Select part that will be hidden
        </h3>,
        <TextHightlight
          className="mb30"
          classNameHighlighted="text-highlight"
          onHighlight={[Function]}
        >
          
        </TextHightlight>,
        <h3
          className="mb5"
        >
          3. Question preview
        </h3>,
        <QuestionPhrase
          disabled={true}
          error=""
          hideIfNotAvailable={true}
          phrase=""
          word="School"
          wordID="school"
        />,
        undefined,
        <div
          className="mb25 mt35"
        >
          <RaisedButton
            disabled={false}
            fullWidth={false}
            label="Save phrase"
            labelPosition="after"
            onClick={[Function]}
            primary={true}
            secondary={false}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "1. Set phrase",
        },
        "ref": null,
        "rendered": "1. Set phrase",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "mb25",
          "disabled": false,
          "errorText": undefined,
          "floatingLabelFixed": false,
          "floatingLabelText": "Phrase",
          "fullWidth": false,
          "multiLine": false,
          "onChange": [Function],
          "rows": 1,
          "style": Object {
            "width": "100%",
          },
          "type": "text",
          "underlineShow": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "2. Select part that will be hidden",
          "className": "mb5",
        },
        "ref": null,
        "rendered": "2. Select part that will be hidden",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "",
          "className": "mb30",
          "classNameHighlighted": "text-highlight",
          "end": undefined,
          "onHighlight": [Function],
          "start": undefined,
        },
        "ref": null,
        "rendered": "",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "3. Question preview",
          "className": "mb5",
        },
        "ref": null,
        "rendered": "3. Question preview",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": true,
          "endIndex": undefined,
          "error": "",
          "hideIfNotAvailable": true,
          "phrase": "",
          "startIndex": undefined,
          "word": "School",
          "wordID": "school",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <RaisedButton
            disabled={false}
            fullWidth={false}
            label="Save phrase"
            labelPosition="after"
            onClick={[Function]}
            primary={true}
            secondary={false}
          />,
          "className": "mb25 mt35",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "fullWidth": false,
            "label": "Save phrase",
            "labelPosition": "after",
            "onClick": [Function],
            "primary": true,
            "secondary": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h3>
            1. Set phrase
          </h3>,
          <TextField
            className="mb25"
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelText="Phrase"
            fullWidth={false}
            multiLine={false}
            onChange={[Function]}
            rows={1}
            style={
              Object {
                "width": "100%",
              }
            }
            type="text"
            underlineShow={true}
            value=""
          />,
          <h3
            className="mb5"
          >
            2. Select part that will be hidden
          </h3>,
          <TextHightlight
            className="mb30"
            classNameHighlighted="text-highlight"
            onHighlight={[Function]}
          >
            
          </TextHightlight>,
          <h3
            className="mb5"
          >
            3. Question preview
          </h3>,
          <QuestionPhrase
            disabled={true}
            error=""
            hideIfNotAvailable={true}
            phrase=""
            word="School"
            wordID="school"
          />,
          undefined,
          <div
            className="mb25 mt35"
          >
            <RaisedButton
              disabled={false}
              fullWidth={false}
              label="Save phrase"
              labelPosition="after"
              onClick={[Function]}
              primary={true}
              secondary={false}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "1. Set phrase",
          },
          "ref": null,
          "rendered": "1. Set phrase",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "mb25",
            "disabled": false,
            "errorText": undefined,
            "floatingLabelFixed": false,
            "floatingLabelText": "Phrase",
            "fullWidth": false,
            "multiLine": false,
            "onChange": [Function],
            "rows": 1,
            "style": Object {
              "width": "100%",
            },
            "type": "text",
            "underlineShow": true,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "2. Select part that will be hidden",
            "className": "mb5",
          },
          "ref": null,
          "rendered": "2. Select part that will be hidden",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "",
            "className": "mb30",
            "classNameHighlighted": "text-highlight",
            "end": undefined,
            "onHighlight": [Function],
            "start": undefined,
          },
          "ref": null,
          "rendered": "",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "3. Question preview",
            "className": "mb5",
          },
          "ref": null,
          "rendered": "3. Question preview",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": true,
            "endIndex": undefined,
            "error": "",
            "hideIfNotAvailable": true,
            "phrase": "",
            "startIndex": undefined,
            "word": "School",
            "wordID": "school",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <RaisedButton
              disabled={false}
              fullWidth={false}
              label="Save phrase"
              labelPosition="after"
              onClick={[Function]}
              primary={true}
              secondary={false}
            />,
            "className": "mb25 mt35",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "disabled": false,
              "fullWidth": false,
              "label": "Save phrase",
              "labelPosition": "after",
              "onClick": [Function],
              "primary": true,
              "secondary": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`renders correctly with phrase data passed 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PhraseWordSettings
    onSubmit={[MockFunction]}
    phrase={
      Object {
        "endIndex": 10,
        "phrase": "Lorem ipsum",
        "startIndex": 0,
      }
    }
    submitText="Save phrase"
    word="School"
    wordID="school"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h3>
          1. Set phrase
        </h3>,
        <TextField
          className="mb25"
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelText="Phrase"
          fullWidth={false}
          multiLine={false}
          onChange={[Function]}
          rows={1}
          style={
            Object {
              "width": "100%",
            }
          }
          type="text"
          underlineShow={true}
          value="Lorem ipsum"
        />,
        <h3
          className="mb5"
        >
          2. Select part that will be hidden
        </h3>,
        <TextHightlight
          className="mb30"
          classNameHighlighted="text-highlight"
          end={10}
          onHighlight={[Function]}
          start={0}
        >
          Lorem ipsum
        </TextHightlight>,
        <h3
          className="mb5"
        >
          3. Question preview
        </h3>,
        <QuestionPhrase
          disabled={true}
          endIndex={10}
          error=""
          hideIfNotAvailable={true}
          phrase="Lorem ipsum"
          startIndex={0}
          word="School"
          wordID="school"
        />,
        undefined,
        <div
          className="mb25 mt35"
        >
          <RaisedButton
            disabled={false}
            fullWidth={false}
            label="Save phrase"
            labelPosition="after"
            onClick={[Function]}
            primary={true}
            secondary={false}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "1. Set phrase",
        },
        "ref": null,
        "rendered": "1. Set phrase",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "mb25",
          "disabled": false,
          "errorText": undefined,
          "floatingLabelFixed": false,
          "floatingLabelText": "Phrase",
          "fullWidth": false,
          "multiLine": false,
          "onChange": [Function],
          "rows": 1,
          "style": Object {
            "width": "100%",
          },
          "type": "text",
          "underlineShow": true,
          "value": "Lorem ipsum",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "2. Select part that will be hidden",
          "className": "mb5",
        },
        "ref": null,
        "rendered": "2. Select part that will be hidden",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Lorem ipsum",
          "className": "mb30",
          "classNameHighlighted": "text-highlight",
          "end": 10,
          "onHighlight": [Function],
          "start": 0,
        },
        "ref": null,
        "rendered": "Lorem ipsum",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "3. Question preview",
          "className": "mb5",
        },
        "ref": null,
        "rendered": "3. Question preview",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": true,
          "endIndex": 10,
          "error": "",
          "hideIfNotAvailable": true,
          "phrase": "Lorem ipsum",
          "startIndex": 0,
          "word": "School",
          "wordID": "school",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <RaisedButton
            disabled={false}
            fullWidth={false}
            label="Save phrase"
            labelPosition="after"
            onClick={[Function]}
            primary={true}
            secondary={false}
          />,
          "className": "mb25 mt35",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "fullWidth": false,
            "label": "Save phrase",
            "labelPosition": "after",
            "onClick": [Function],
            "primary": true,
            "secondary": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h3>
            1. Set phrase
          </h3>,
          <TextField
            className="mb25"
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelText="Phrase"
            fullWidth={false}
            multiLine={false}
            onChange={[Function]}
            rows={1}
            style={
              Object {
                "width": "100%",
              }
            }
            type="text"
            underlineShow={true}
            value="Lorem ipsum"
          />,
          <h3
            className="mb5"
          >
            2. Select part that will be hidden
          </h3>,
          <TextHightlight
            className="mb30"
            classNameHighlighted="text-highlight"
            end={10}
            onHighlight={[Function]}
            start={0}
          >
            Lorem ipsum
          </TextHightlight>,
          <h3
            className="mb5"
          >
            3. Question preview
          </h3>,
          <QuestionPhrase
            disabled={true}
            endIndex={10}
            error=""
            hideIfNotAvailable={true}
            phrase="Lorem ipsum"
            startIndex={0}
            word="School"
            wordID="school"
          />,
          undefined,
          <div
            className="mb25 mt35"
          >
            <RaisedButton
              disabled={false}
              fullWidth={false}
              label="Save phrase"
              labelPosition="after"
              onClick={[Function]}
              primary={true}
              secondary={false}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "1. Set phrase",
          },
          "ref": null,
          "rendered": "1. Set phrase",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "mb25",
            "disabled": false,
            "errorText": undefined,
            "floatingLabelFixed": false,
            "floatingLabelText": "Phrase",
            "fullWidth": false,
            "multiLine": false,
            "onChange": [Function],
            "rows": 1,
            "style": Object {
              "width": "100%",
            },
            "type": "text",
            "underlineShow": true,
            "value": "Lorem ipsum",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "2. Select part that will be hidden",
            "className": "mb5",
          },
          "ref": null,
          "rendered": "2. Select part that will be hidden",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Lorem ipsum",
            "className": "mb30",
            "classNameHighlighted": "text-highlight",
            "end": 10,
            "onHighlight": [Function],
            "start": 0,
          },
          "ref": null,
          "rendered": "Lorem ipsum",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "3. Question preview",
            "className": "mb5",
          },
          "ref": null,
          "rendered": "3. Question preview",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": true,
            "endIndex": 10,
            "error": "",
            "hideIfNotAvailable": true,
            "phrase": "Lorem ipsum",
            "startIndex": 0,
            "word": "School",
            "wordID": "school",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <RaisedButton
              disabled={false}
              fullWidth={false}
              label="Save phrase"
              labelPosition="after"
              onClick={[Function]}
              primary={true}
              secondary={false}
            />,
            "className": "mb25 mt35",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "disabled": false,
              "fullWidth": false,
              "label": "Save phrase",
              "labelPosition": "after",
              "onClick": [Function],
              "primary": true,
              "secondary": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
